ostree_upload_server/importers/tar.py

```diff
@@ -0,0 +1,80 @@
+import logging
+import tempfile
+import tarfile
+
+import urllib
+import urlparse
+
+from gi.repository import GLib
+
+from os import path
+from shutil import rmtree
+
+from .base import BaseImporter
+from .util import find_repo, open_repository
+
+class TarImporter(BaseImporter):
+    MIME_TYPE = 'application/x-tar'
+
+    def _import_commit(self, target_repo, src_path_obj, refs=None):
+        if not self._source_repo_path:
+            raise RuntimeError("Cannot invoke _import_commit without calling import_to_repo first")
+
+        if not refs:
+            raise RuntimeError("Cannot invoke _import_commit without valid refs")
+
+        logging.debug("Importing %s%s into %s...", self._source_repo_path,
+                      refs,
+                      target_repo.get_path().get_uri())
+
+        options = GLib.Variant('a{sv}', {
+            'refs': GLib.Variant('as', refs),
+            'inherit-transaction': GLib.Variant('b', True),
+        })
+
+        source_repo_uri = urlparse.urljoin('file:', urllib.pathname2url(self._source_repo_path))
+        target_repo.pull_with_options(source_repo_uri,
+                                      options,
+                                      None, None)
+
+        logging.info("Importing complete.")
+
+    def import_to_repo(self):
+        logging.info('Trying to use %s extractor...', self.__class__.__name__)
+
+        dest_path = tempfile.mkdtemp('', self.__class__.__name__)
+        try:
+            logging.info('Extracting \'%s\' to a temp dir in %s...', self._src_path, dest_path)
+            with tarfile.open(self._src_path) as tar_archive:
+                tar_archive.extractall(path=dest_path)
+
+            logging.info('Applying delta from \'%s\' to \'%s\'')
+
+            self._source_repo_path = find_repo(dest_path)
+            source_repo = open_repository(self._source_repo_path)
+
+            refs = source_repo.list_refs().out_all_refs
+            logging.debug("Refs: %r", refs)
+
+            if not refs:
+                logging.error("Could not find any refs in the source repo!")
+                raise RuntimeError("Missing refs in Tar ostree repository!")
+
+            # We only process the first ref in the repo provided!
+            ref = refs.keys()[0]
```

Comment:
If you're just relying on the fact that the build will only create one relevant ref, then I think you should raise an exception if there's more than one ref so we don't randomly import the wrong thing.

Reply:
Fair enough.